simularium/simularium-viewer
# Caching: refactor, memory improvements, trimming

Time Estimate or Size
=======
_large: many lines that touch many important parts of the viewer, but some of the large line count is tests!_

Problem
=======
Closes #372 and also refactors to address memory and performance concerns, and lay the groundwork for future updates to caching.

paired with @toloudis 

Solution
========
Initial attempts to trim the cache during playback encountered performance and memory bottlenecks that dramatically slowed down the viewer. These seem to be primarily related to triggering an excess of garbage collection events. Poking around on memory profiling revealed several places in the code that data was being duplicated, particularly copies of `AgentData`. The result of these explorations is a refactored cache that has a number of other changes accompanying it, both to provide the new functionality, and to address these memory and performance concerns.

### Linked list cache
The cache is now it's own class in `VisDataCache.ts`. It takes an optional settings prop:
```
interface VisDataCacheSettings {
    maxSize: number;
    cacheEnabled: boolean;
}
```
and provides a public method, `changeSettings` to update those properties after initializing.

Like before, if the cache is disabled, only the current frame is ever stored.

When the cache is enabled and a `maxSize` has been provided, the cache will be trimmed before adding any new frame to keep its size under the limit. The trimming behavior could probably be made more sophisticated in the future.

Refactoring the cache to be a linked list of `CachedFrames` was effective in reducing the garbage collection events that seemed to be slowing things down. Maybe a matter of opinion but I also find that refactoring it into its own file and class makes working with cache methods easier than the previous two arrays of data and metadata. 

For the future: this cache format can be refactored to facilitate a multi-part cache with non-continuous sections of cached frames, which could reduce load on the back end by preserving local cached data whenever possible.

### Move parsing of `AgentData` to `updateScene`
We have moved the parsing of `AgentData` from the functions that handle incoming binary and JSON data to the call of `updateScene` in `VisGeometry`.
This makes `updateScene` more expensive, but does not increase the number of times data needs to be parsed.
By caching the raw incoming data largely unparsed (except for critical metadata like time, frameNumber, agentCount) we reduce the memory footprint of the cached data, and the processing needed to cache the frames.
### Object pooling of `VisAgentInstances`
Some trajectories efficiently reuse agent Ids and others do not. In cases where the number of unique Ids climbs throughout playback, the `VisAgentInstances` would get larger and larger, holding excess copies of stale `AgentData`
We now only store current agent instances. The unused `VisAgent` objects released by old agents are pooled in an array and re-used, rather than instantiating new ones every time. In cases where it matter this has a dramatic effect on memory use.

## Type of change
Please delete options that are not relevant.
* New feature (non-breaking change which adds functionality)
* This change requires updated or new tests


Steps to Verify:
----------------
1. In the test bed load and run binary trajectories with different settings (cache enabled/not, cache enabled with and without maxSize provided), confirm that everything runs smoothly with unlimited cache and with trimming (tip: adding a log in `trimCache` to confirm that you have hit the limit and begun trimming makes this easier)
2. Locally load A JSON trajectory and confirm they it loads properly correctly.
3. Installing a dev build into the website is also a nice way to confirm things aren't busted.
